ibis/expr/datatypes/value.py

```diff
@@ -37,6 +38,16 @@ def infer(value: Any) -> dt.DataType:
     )
 
 
+@infer.register(Deferred)
+def infer_deferred(value: Deferred) -> NoReturn:
```

Comment:
sorry for my ignorance: is there somewhere I can refer to to get an idea of circumstances that cause inference to be triggered?  seems like its not when building a Deferred which makes sense, so presumably its when incorporating a Deferred into an expression bound to a concrete table or column?

Reply:
honestly, I don't see/understand all the cases either.

This is one example that exercises it:

```python
import ibis

t = ibis.memtable({"a": [1, 2, 3], "b": [4, None, 6]})
t.fill_null({"a": ibis._.b})
```

I found this by looking up call sites of dt.infer().

I'm not sure this is required/correct to use infer() here. I think probably we should actually be binding the values to the table first, and then we would be able to resolve their values.

Reply:
So maybe we should actually frame this error message as "this is unexpected behavior, please file a bug with the ibis maintainers"?